### ISSUE TITLE
Issues with localization/translations.

### ISSUE THREAD

**Issue by bummzack** [burnbright/silverstripe-shop#368]
There are some issues regarding localization of this module (I'm using 1.1.0).

 - [x] The translation files are mostly still in the old format (php files instead of yml). Yet using the legacy-adapter for the PHP files doesn't work (because `i18n::include_locale_file` doesn't exist). 

 - [ ] Existing YML files are incomplete (`en.yml`, `nn.yml`) or broken (`es_ES.yml` should be named `es.yml`, `nb.yml` throws a formatting error)

 - [x] Templates use the old syntax for translations (`<% _t('KEY', 'Default value') %>` instead of `<%t KEY 'Default value' %>`).

 - [x] Translations use sprintf (sequential) placeholder-replacement instead of the new variable-based replacement.

 - [x] Translation keys don't exist or are improperly linked. See **Example 1**.

 - [x] There are several duplicate translations. Instead of having `OrderInformation.ss.QUANTITY`, `OrderInformation_Editable.ss.QUANTITY`, `OrderInformation_PackingSlip.ss.QUANTITY` and `ProductCategoryItem.ss.QUANTITY` it would probably be sufficient to have `Order.QUANTITY` and use this identifier in several templates. There are also translations for `Order_statusEmail.ss'` and `Order_StatusEmail.ss'` in some of the lang-files, the first one is probably a typo (or legacy value) and can be deleted.

#### Example 1:

The template File [`Cart.ss`](https://github.com/burnbright/silverstripe-shop/blob/1.1.0/templates/Includes/Cart.ss) contains several translation that don't exist. Excerpt:

```html
<th scope="col"><% _t("PRODUCT","Product") %></th>
<th scope="col"><% _t("UNITPRICE","Unit Price") %></th>
<th scope="col"><% _t("QUANTITY", "Quantity") %></th>
<th scope="col"><% _t("TOTALPRICE","Total Price") %> ($Currency)</th>
```

There is no translation for `Cart.ss.PRODUCT`, nor for any of the other of these values. Instead the default value from the template is being used. One should be able to replace all the placeholder values with "Translation missing" and still never see that string!

----

I guess this is a known issue. What would be a good way to go about this? I guess starting with a proper language file for english (`en.yml`) should be the first step?

**Comment by markguinn** [comment]
Thanks for raising this. We're planning to spend some time on this subject at the Silverstripe Europe meetup in a few weeks' time. Translation is not something I have a lot of experience with personally though. Are you just raising the issue or do you think you'll have some time to contribute to working on this (either is fine of course)?

**Comment by bummzack** [comment]
I'll hopefully have some time to work on this issue. There are a lot of files that would be modified in the process of resolving the issue though... Templates, code and of course the language files themselves.

I noticed that the templates also use the old variable syntax, eg. `<% if MyVar %>` instead of `<% if $MyVar %>`. This could also be updated in the process of changing the `_t` calls in the templates.

I'd probably go about this as follows:

 1. Collect all translatable strings (maybe with the `i18nTextCollectorTask`)
 2. Build a clean `en.yml` file with the collected strings. Remove duplicates and use speaking keys/identifiers. Use `{variable}` placeholders instead of sprintf placeholders like `%s`, `%d` etc.
 3. Update code/templates to use the proper translation keys/variables.
 4. Try to create translation files for the other languages, by re-using the values that are already present in the `.php` files.. although I imagine this to be rather difficult for a non native-speaker.. so these files should be created or at least be reviewed by a native-speaker.

What do you think?

What are your plans for the Europe-Meetup btw.? Do you want to ask the community about best-practice or help? I'll be there as well, looking forward to meet you ;)

**Comment by markguinn** [comment]
This sounds really good to me. I think step 4 could be tedious but doable and maybe we can recruit some help at the meetup. I wonder if there's an automated solution laying around as well from when the main SS modules went through the same transition.

I'm hoping to spend part of the session giving an overview of the different options for ecommerce on Silverstripe, as well as my vision for ss-shop as one of those options. But I do want to spend a good portion of the session hearing what other people are doing, as well as chatting collaboratively about how to make commerce on Silverstripe less painful overall and what it would take to have an ecommerce module that's high quality and well supported in the long run. Internationalization would be a big part of that in my view. Looking forward to meeting you as well!

**Comment by bummzack** [comment]
Sounds interesting!

I'll probably tackle this issue pretty soon, since I'll be using the shop-module for a non-english website. 

Ingo Schommer wrote a [task](https://github.com/chillu/i18n_yml_converter) that converts the legacy PHP files to YML files. From my experience there's still a lot of manual changes required to get usable language files though, so just running the existing PHP files through a converter might not be the best step towards a better solution.

**Comment by markguinn** [comment]
@bummzack you've used both SwipeStripe and SS Shop right? I'd love to get your input on the pros and cons of each. If you have time and interest you could send me some thoughts via email (mark@adaircreative.com)?

**Comment by bummzack** [comment]
I have started tackling this issue.
There are lots of values that are missing in the locale files (strings are only defined in code/templates, but not listed in locale-files).

This will make it very hard to re-use the existing translation-strings. I'll have to ditch all the translation files and build a clean file for `en`. It then has to be a community effort to create (new) translations based on the `en` file. I'll try to supply a `de` locale-file as well.

**Comment by markguinn** [comment]
Thanks for taking this on! I trust your judgment. I'm going to ask Anselm to comment on this too because he just worked on a German site using ss-shop.

**Comment by bummzack** [comment]
Oh. I'd really be interested how he did that. From what I can tell, the translations are completely broken and won't work in any language other than english. Pinging @anselmdk :)

**Comment by anselmdk** [comment]
@bummzack I used @zauberfisch's module https://github.com/Zauberfisch/silverstripe-better-i18n to create the `en.yml` and had it translated to German. That way there's no need for Transifex (which I never get to wrap my head around, and which seems too much bloat to me). But what do I know...
Here are the files:
https://gist.github.com/anselmdk/4f25068b5e1c80305de2

**Comment by bummzack** [comment]
@anselmdk thanks for the swift reply. I assumed you did something along these lines. Also thanks a lot for sharing your german translations. This will be a great starting point for the `de.yml` I'm planning to provide.

I won't be using it *as is* though, since my goal is also to steer away from the `sprintf` variable injection and use proper template syntax (as outlined in the official docs https://docs.silverstripe.org/en/3.2/developer_guides/i18n/)

**Comment by anselmdk** [comment]
Np, happy to help. There's a lot of redundant strings in there though...
We were discussing using Transifex, but I somehow feel that's a little overkill, and wouldn't know how to go ahead with it. @nobrainerweb what's your opinion on this, and what would you do?

**Comment by bummzack** [comment]
I think it would probably be worthwhile to add this project to transifex. I don't have experience with it either, but I think you'll have to provide at least a base (eg. english) there?

By solving this issue here I'm planning to provide an up-to-date english language file where I try to remove redundancies and also update the used syntax. I assume this file can then be used as a basis to feed into transifex?

**Comment by markguinn** [comment]
Yes, I don't have a ton of experience with Transifex either but my understanding is that any existing file can be fed in as a base and then other people can add to it through Transifex instead of a git checkout and a pull request. I'll see if I can find someone on IRC who has experience with it and get some wisdom but I don't think it needs to hold you up @bummzack.

**Comment by bummzack** [comment]
What do you guys think about using `db_<FieldName>` for field-label translations? 

Example: `Address` has a field called `Country`. The translation key is `Address.COUNTRY`. We could create a translation key `Address.db_Country` instead. The benefit of this is, that it will automatically localize field-labels when somebody uses form-scaffolding.

Since this issue will alter so many files in a non-backwards compatible way, I think we could do this just as well…

**Comment by markguinn** [comment]
In what way will this break existing shops (say, the one Anselm just built for example)? Ideally this could go in 1.2, but I think we need to think carefully how this affects semantic versioning. 

If, for example, Anselm has his custom translations working and then he upgrades and suddenly the translation keys all change and his custom translations are no longer applied, that seems to necessitate going into 2.0. Is there any way to prevent that from happening and still allow you to pursue the work you're doing?

**Comment by bummzack** [comment]
It's not an easy thing to answer. In order to do this properly, some translation keys (and existing string values) have to change. Ideally, some would also be merged to remove redundancies.

Now let's assume the following: The translation keys are cleaned up and all calls to `_t` (be it in templates or php) use the appropriate syntax. If there's an updated `de.yml` that's shipping with the update, a site such as Anselm's would not break. If there are people who did the same as Anselm (collect all translatable strings and create a translation-file from scratch), then these sites will break, unless we ship a language file for the language in question.

If somebody created custom templates for the shop and used the translations provided (eg. by writing something like `<% _t("Address.COUNTRY","Country") %>` in his own templates), then these might stop working. But translation-files are incomplete since 2012 (#73). The `php` ones no longer work and some of the exising `yml` ones have syntax errors, or aren't referenced correctly... so chances are pretty low that somebody made use of the existing translations.

I think the deciding factor here is if the language files are considered to be part of a *private* or *public* API. If they are public API, then this change would have to go into `2.0`. If they are private API, then this change could go into `1.2`.

Not sure how changes to translation files are treated in SilverStripe itself? Maybe @tractorcow can help here? 

**TL/DR:** This change can only make it into `1.2` if translations (and their keys) are considered to be private API.

**Comment by markguinn** [comment]
Excellent analysis. I agree on the distinction. I suspect most people who use Shop at this point are either using it for English-only sites or are having to customise templates to the point that they would not be affected by this change. I'm interested to hear what others think though. I've asked this Stack Overflow question as well to get a wider opinion: http://stackoverflow.com/questions/33245058/are-i18n-string-keys-considered-to-be-public-api-for-the-purposes-of-semantic

I lean towards seeing i18n keys as part of the public API but I'm hoping someone can convince me otherwise. :) If we decide it does count I'll be inclined to push 2.0 out sooner and let 3.0 be our larger, SS 4.0-compatible release. I think this is an important change.

**Comment by bummzack** [comment]
Sounds good. Upvoted your question on SO, hope we get some good answers there.

**Comment by tractorcow** [comment]
`db_Field` would be fine for field labels, and it wouldn't require a new major version under semver as long as the old keys were maintained concurrently (perhaps retired at the 2.0 mark).

Transifex also makes life simpler, if you can do it. I have been developing `cow` for use with transifex to speed up module translation. ;D https://github.com/tractorcow/cow/#module-level-commands

As long as you use 'merge=1' with i18nTextCollectorTask, then you won't lose your legacy strings.

I suggest looking at migrating your existing strings directly to the new yml syntax, setting up with transifex, and then running `cow module:translate` over it to refresh your keys from your source code.

**Comment by bummzack** [comment]
@tractorcow Thanks for chiming in. I was also thinking about keeping the old keys for legacy reasons, but what about making use of named variables (`{placeholder}`) instead of sequential replacement (eg. using `sprintf`) in strings?

Tanks for the link to your module. I'll have a look at it.

**Comment by bummzack** [comment]
Thinking about it: I guess it would be fine to create new keys for translations that use the new syntax and keep the ones with old syntax as legacy keys as well.

The main problem I see with the legacy-keys is, that there's probably no easy way to deprecate them? Removing them upon 2.0 release won't trigger any notices. Or is there some feature like this built into the i18n system?

**Comment by tractorcow** [comment]
The best way to deprecate them is to keep a separate "master" en_2.0.yml list.

Once you decide to deprecate the old strings, replace en.yml with this, and force push up to transifex. All old strings will be deleted, and next you pull down translations all .yml files will have deprecated keys removed.

Which is why transifex makes things easier. :)

**Comment by bummzack** [comment]
Let's go forward with this. Here's my proposition:

#### Version 1.2
We can create a complete `en` and `de` file for `1.2` (this is actually rather easy, since we can just use the file provided by Anselm for DE and the one the text-collector generates for EN). Cleaning up the langauge files and updating templates/code to the new syntax is out of scope for `1.2`, since it will break the existing "API". The old PHP files can be converted to YML and be supplied "as is". Although the overlap of the old translations and what's currently being used in code is really small…
 
Users that went the same way as Anselm to create a translation will be able to contribute them as PR into `1.x`. @anselmdk would you be willing to provide your german translation if we go that route?

#### Version 2.0
Do what has been outlined in this issue. Create a clean file for `en`, update `_t` syntax in code and templates. 

#### Transifex
I strongly suggest using transifex, but I'd do that for `2.x` only. Starting with a clean base is much more encouraging to people who are willing to provide translations (reduced duplicates, new syntax etc.).

**Comment by NobrainerWeb** [comment]
+1 for Transifex, let's stick to the stuff that people already know and use.

**Comment by markguinn** [comment]
This all sounds really good to me, Roman. I'm tempted to want to push for Transifex earlier, but there's probably wisdom in your approach. If, for example, we end up making drastic changes for 2.0 as a result of the user stories that will affect the translation keys yet again and if we split the module up (e.g. cart/checkout and product/catalogue) that would also affect how Transifex is set up.

I'll keep reading up on Transifex, though, and if it seems feasible to get started with that earlier maybe do something like tractorcow suggested.

Just out of curiousity - approximately how long would it take to translate on of these files from scratch into a language you speak fluently? Anselm's de.yml file has about 500 lines.

**Comment by bummzack** [comment]
I dont' know how much time goes into translating such a file. Maybe @anselmdk can tell us how much time it took to create the file for German?

Sadly I have no experience with Transifex, so I can't tell how flexible it is in this regard. Maybe somebody with experience (@tractorcow ?) can chime in and answer this quesion:

Assuming we have a file like this (hypothetical):
```yml
en:
  Order:
    SUCCESSFUL: 'Successful'
    CANCELORDER: 'Cancel order %s'
    AMOUNT: 'Amount'
  Order.ss:
    AMOUNT: 'Amount'
    INCOMPLETE: 'Incomplete'
```

And we would merge/improve it into something like this:
```yml
en:
  Order:
    SUCCESSFUL: 'Successful'
    CANCEL_ORDER: 'Cancel order {OrderNumber}'
    AMOUNT: 'Amount'
    INCOMPLETE: 'Incomplete'
```

So, removing duplicates (`AMOUNT`), update variable-replacement syntax (`'Cancel order %s'` => `'Cancel order {OrderNumber}'`) and maybe make keys more readable: `CANCELORDER` => `CANCEL_ORDER`. 

Now if we did this and pushed it to transifex, my fear is, that values like `"Incomplete"` would have to be re-translated, as it was previously under `Order.ss`. Or does Transifex provide some tools to easily re-arrange/assign keys?

**Comment by anselmdk** [comment]
I'll be happy to create a pull request with the English and German files (or the German only). @bummzack / @markguinn I guess I could just go ahead with it?
I like the approach outlined here, and agree that cleaning up the translations would make most sense until after we've agreed on if we want to restructure / split up the module.

**Comment by bummzack** [comment]
@anselmdk Yeah, I think creating a PR is fine. Did you merge your translations with the values from the legacy files or just used the output of the zauberfisch-collector? Removing the legacy keys can break somebodys application if he used translations in his own templates which are no longer being used in the shop-module itself. That's why I'm pointing it out.

**Comment by markguinn** [comment]
From what I've read it's safe to add keys but we can't remove them until 2.0. It's also a good idea to change the key if the contents changes (e.g. converting from sprintf to variable syntax). @bummzack I think we might be able to to fit more of this into 1.2 than you think. Perhaps something like:

* For 1.2:
  * convert from legacy php syntax to yml files and make sure they're named correctly
  * create new keys for variable syntax and make sure they're used in the code and templates
  * for duplicates, pick one of the two and make sure it's used exclusively in code and templates
  * if there are are keys you feel are named incorrectly, you could add them but i'd encourage you to keep these to a minimum
  * make a list of the deprecated keys (I think you can have comments in the yml files for this too)
* For 2.0:
  * remove deprecated keys

Am I missing something there? The downside is we're starting from scratch with translations, but it sounds like we might have to do that anyway? I don't want to overcomplicate things.

**Comment by bummzack** [comment]
I just created a PR (#378) where I converted all PHP files to YML and merged them with the existing YML files (if applicable).

This is just a housekeeping things. It doesn't improve the state of the translations in general, just that we can now deal with YML files instead of PHP files.

The next step would be to merge the actual text content (eg. the files produced by the textcollector) into these files. With this procedure we could generate a complete translation file for `en` and `de`, that contains all legacy keys/values as well.

@markguinn If we add new keys and use them within the code, wouldn't that be violating the public API as well? If somebody did the same as Anselm and built a translations-file using the text-collector, the new keys won't be translated in such a file. As far as I can tell, pretty much any changes to translation keys or format (eg. sprintf vs. variables) can break things.

**Comment by bummzack** [comment]
Just as a note to anybody who's going to want to merge YML files. I used the procedure outlined in this [StackOverflow Question](http://stackoverflow.com/questions/25630633/merging-yaml-config-files-recursively-with-bash).

Files where merged so that the newer file overrides existing keys from the older file.

**Comment by markguinn** [comment]
That's a tough call. @tractorcow how would you treat such a change in Silverstripe core (TL;DR; = Adding a new translation key and converting all instances of an older key in the code and default templates to use the new key)?

Isn't this essentially analogous to change the text of a button? When I think of it that way it doesn't seem like breaking public API to me.

**Comment by bummzack** [comment]
Here comes another PR, this time with the files from Anselm merged in: #379

The `en` and `de` YML files should now contain all translatable keys from the shop-module. Having a look at the shop with locale set to German shows still several untranslated text-fragments, but these aren't wrapped in `_t` calls ;-)

@anselmdk I assume that shop isn't live yet? Your yml file seems to contain some syntax errors ([L160](https://gist.github.com/anselmdk/4f25068b5e1c80305de2#file-de-yml-L160), [L206](https://gist.github.com/anselmdk/4f25068b5e1c80305de2#file-de-yml-L206), etc.). I have corrected these as good as possible before the merge. Maybe this is due to a bug in the zauberfisch text-collector?

**Comment by anselmdk** [comment]
@bummzack no I'm not using it in production, so that's perfect - great work!

**Comment by markguinn** [comment]
Is it true that the first 3 todo items in the original issue are done?

**Comment by bummzack** [comment]
@markguinn Only the first item is done. Will update that. The existing YML files are still incomplete and template syntax has not been updated yet.

I'd rather work with broad strokes here and aim for a 2.0 release for the changes. There's so much that has to be done for proper translation of this module.

**Comment by markguinn** [comment]
No hurry, I was just trying to get a feel for where we're at. Thanks for all your work on this!

**Comment by bummzack** [comment]
@markguinn What do you think about a development branch for 2.0? I might split this into several PRs and then you can merge them into the 2.0 development branch?

**Comment by bummzack** [comment]
Here's another step towards a fully translated shop-module: #381 

While this doesn't improve the template-syntax or declutters the YAML files, it aims to make `1.2` translatable without violating semver constraints.

**Comment by markguinn** [comment]
I'll have a look at this in the morning, Roman. Thanks. I think it probably does make sense to start a 2.0 branch soon. I've just been thinking through the best way to do that with gitflow.

**Comment by markguinn** [comment]
I just created a branch called `develop-2.0`. Any breaking changes can be pushed there and we can keep it up to date with the main `develop` branch.

**Comment by bummzack** [comment]
I'll start with this now. I've noticed that there seems to be a mixture of translation key formats. Some are `ALL_CAPS`, some are `CamelCase` etc.

It seems, this also isn't consistent in the SilverStripe Framework itself, newer additions seem to favor `CamelCase` keys ([example](https://github.com/silverstripe/silverstripe-framework/blob/3/lang/en.yml#L217))?

I'm going to change/merge a lot of the translation keys now, so please let me know which format you would prefer?

**Comment by markguinn** [comment]
I think we should always stick with silverstripe when possible. Camel case is good with me.

**Comment by bummzack** [comment]
Ok, there's probably always going to be a mixture, because there are standard-keys such as `SINGULARNAME` etc. But I'm going to change everything that's under our control to camel case then.

**Comment by bummzack** [comment]
I made some progress on this, see #408

I didn't delete the existing language files, but as of now, only the `en.yml` file is in a usable state. I think the best way forward now would be to remove the existing translations and start using Transifex with the new `en.yml` file as basis for fresh translations.

**BUT**: I'd hope you guys could have a look at the `en.yml` file, maybe you'll find some things that could be merged/combined further or that could be improved before moving to Transifex.

There are still some translations in the codebase that probably only work well with western languages, such as this one:
```
<span class="discounted">$DiscountedPrice.Nice</span> <%t PriceTag.SAVE "Save" %>: <span class="savings">$DiscountedPrice.Savings</span>
```

It would probably be better to use translations and placeholders for this, something like:

```
<%t PriceTag.Save "{DiscountedPrice} Save: {Savings}" DiscountedPrice=$DiscountedPrice.Nice Savings=$DiscountedPrice.Savings %>
```

But then the markup (`<span>`) is gone and I kinda dislike having markup in the translations? Any opinions on this?

**Comment by anselmdk** [comment]
Looks good. Regarding discounted price, I feel your pain, but I see we've already got it with `Checkout.TermsAndConditionsLink`. I'd say let's go ahead with it.

**Comment by bummzack** [comment]
I've reconsidered my position regarding the example above (PriceTag). From a developer point of view it's really simple to create a custom template, eg. a custom template for `PriceTag`. Having all the markup in the translations is going to somewhat limiting in that regard. Especially translations that are used just once and in templates that only consists of a few lines of markup, I think it's actually better to keep the translations as "small grained" as possible and add the markup in the template.

**Comment by bummzack** [comment]
If there aren't any objections, I'm going to close this one and open a new issue for Transifex and translating to the different locales.

What should we do with the existing language files? If I move them to a subfolder of `lang` will they no longer be detected by the i18n? That would be ideal.

**Comment by markguinn** [comment]
I agree about the PriceTag. In my experience we end up overriding most of the stock templates anyway.

I don't know the answer about the old lang files. It's fine with me if you create a "legacy-lang" folder (or whatever you want to call it) for them if nesting doesn't work.

I did create an account on Transifex at one point. I'll dig that up and see if I can give you access.

**Comment by bummzack** [comment]
I'm closing this one. Thanks for your help, everybody.
@markguinn Yeah, please add me to your transifex team. I'm using my github login, so you should find me with the email address you already have.

**Comment by bummzack** [comment]
@tractorcow Can I ask you something more about transifex? We've set up our project there, everything looks good. The translation-interface really lacks some context about the strings that should be translated though… Something like "Name" is really ambiguous. I think a lot of the values would benefit from comments that explain the context.

It seems like comments from the YAML file will get imported into transifex, but will they also survive a i18nTextCollector task? How do you deal with this? Do you add comments via the SilverStripe `_t` API? What about properties like `PLURALNAME`, `MENUTITLE` etc. that can't be annotated? Or do you just manually edit in "instructions" in Transifex?

**Comment by bummzack** [comment]
@tractorcow I guess I figured this one out myself. Generated YAML files won't contain *any* comments. Eg. existing ones will be removed and comments from the i18n API won't show up either. That probably leaves only one option: Manually add comments/instructions within transifex :(

**Comment by markguinn** [comment]
@bummzack is that something the better i18n module (https://github.com/Zauberfisch/silverstripe-better-i18n) might fix? It claims to "merge with existing lang file" which might mean those comments are preserved?

**Comment by bummzack** [comment]
@markguinn Sadly not. Neither the module by @Zauberfisch nor the default `i18nTextCollectorTask` keep comments. They don't even write the comments into the file… I guess it's because the underlying YAML writer doesn't support it at all? (https://github.com/silverstripe/silverstripe-framework/blob/3/i18n/i18nTextCollector.php#L799)

**Comment by tractorcow** [comment]
> It seems like comments from the YAML file will get imported into transifex, but will they also survive a i18nTextCollector task?

We really don't make an attempt to retain those comments anywhere... maybe try out @Zauberfsch's module? 

Maybe just make your keys descriptive?

**Comment by bummzack** [comment]
> We really don't make an attempt to retain those comments anywhere

Yeah, that's understandable, but at least the comments added via i18n API should show up in the generated output… but I guess this is a known issue (the [TODO comment](https://github.com/silverstripe/silverstripe-framework/blob/3/i18n/i18nTextCollector.php#L799) in `i18nTextCollector`seems to indicate that).
